dna/zomes/courses/code/src/course.rs

```diff
@@ -157,82 +261,173 @@ pub fn anchor_entry_def() -> ValidatingEntryType {
+    let anchor_entry = anchor_all_courses_entry();
+    // if this entry already exists, we'll get it's address without creating anything new
+    let anchor_address = hdk::commit_entry(&anchor_entry)?;
+
+    // create a CourseAnchor entry and commit it to DHT so we have it's address
+    let course_anchor_entry =
+        CourseAnchor::new(AGENT_ADDRESS.to_string().into(), timestamp).entry();
+    let course_anchor_address = hdk::commit_entry(&course_anchor_entry)?;
+
+    // link CourseAnchor to anchor_all_courses for this course to be discoverable
+    hdk::link_entries(&anchor_address, &course_anchor_address, "course_list", "")?;
+
+    // create a CourseData entry & commit it to DHT
+    let course_data_entry = CourseData::new(
+        title,
+        AGENT_ADDRESS.to_string().into(),
+        course_anchor_address.clone(),
+        timestamp,
+    )
+    .entry();
+    let course_data_address = hdk::commit_entry(&course_data_entry)?;
+
+    // link CourseAnchor to the CourseData entry for it to be discoverable
+    hdk::link_entries(
+        &course_anchor_address,
+        &course_data_address,
+        "course_anchor->course_data",
+        "",
+    )?;
 
-    Ok(new_course_address)
+    Ok(course_anchor_address)
 }
 
 pub fn update(
     title: String,
     modules_addresses: Vec<Address>,
     course_address: &Address,
 ) -> ZomeApiResult<Address> {
-    let course: Course = hdk::utils::get_as_type(course_address.clone())?;
+    //TODO:  check if addr we receive is actually CourseAnchor, not CourseData
 
-    let new_version_course = Course::from(
+    // get latest CourseData entry by looking at links of CourseAnchor entry
+    let latest_course_data_addr =
+        get_latest_link_addr(course_address, "course_anchor->course_data")?;
+    let course_data: CourseData = hdk::utils::get_as_type(latest_course_data_addr)?;
+
+    // create a new version of CourseData entry
+    let new_course_data_entry = CourseData::from(
         title,
-        course.teacher_address,
-        course.timestamp,
+        course_data.teacher_address,
+        course_data.course_anchor,
+        // TODO: we shouldn't be using old timestamp here. Maybe we should receive it as a parameter for update?
+        course_data.timestamp,
         modules_addresses,
-    );
-    let new_version_course_entry = new_version_course.entry();
+    )
+    .entry();
+    let new_course_data_address = hdk::commit_entry(&new_course_data_entry)?;
 
-    hdk::update_entry(new_version_course_entry, course_address)
+    hdk::link_entries(
+        &course_address,
+        &new_course_data_address,
+        "course_anchor->course_data",
+        "",
+    )?;
+
+    // since CourseAnchor entry stays the same, we're not returning any new addresses here
+    // and since we don't have ownership of the course_address in this method, we're cloning it
+    // to comply with the return type requirements
+    Ok(course_address.clone())
```

Comment:
We can also just return ()

Reply:
Ah, the idea was to comply with the API format returning some address as result. But perhaps with this pattern we could change that (since it is very common to return never changing anchor address).